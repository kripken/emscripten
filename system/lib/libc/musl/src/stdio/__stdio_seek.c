#include "stdio_impl.h"
#ifndef __EMSCRIPTEN__
#include <unistd.h>
#endif

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
#ifdef __EMSCRIPTEN__
	off_t ret;
	if (__wasi_syscall_ret(__wasi_fd_seek(f->fd, off, whence, &ret)))
		ret = -1;
	return ret;
#else
	return __lseek(f->fd, off, whence);
#endif // __EMSCRIPTEN__
}
