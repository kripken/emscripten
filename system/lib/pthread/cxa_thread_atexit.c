/*
 * Copyright 2019 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 * Inspired by libcxxabi/src/cxa_thread_atexit.cpp.
 * The main reasons we don't use that version direclty are:
 * 1. We want to be able to use __cxa_thread_atexit in pure C programs
 *    where libcxxabi is not linked in at all.
 * 2. The libcxxabi relies on TLS variables, which we can't use here.
 *
 * WARNING: We specifically can't use TLS variables or pthread_once in this file
 * because they both end up calling __cxa_thread_atexit.
 *
 * TLS variables: TLS data gets freed using __cxa_thread_atexit.  This means
 * that if we use TLS to store our DtorList it will be free'd while porcessing
 * itself.  See lib/pthread/emscripten_tls_init.c.
 *
 * pthread_once: calls pthread_cleanup_push / pthread_cleanup_pop, which call
 * __cxa_thread_atexit on first use.  See src/thread/pthread_once.c.
 */
#include <assert.h>
#include <stdbool.h>
#include <stdatomic.h>
#include <pthread.h>
#include <stdlib.h>

#include "libc.h"

typedef void (*Dtor)(void*);

typedef struct DtorList {
  Dtor dtor;
  void* obj;
  struct DtorList* next;
} DtorList;

void run_dtors(void* arg) {
  DtorList* dtors = (DtorList*)arg;
  DtorList* head;
  while ((head = dtors)) {
    dtors = head->next;
    head->dtor(head->obj);
    free(head);
  }
}

static pthread_key_t key;
static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

static void ensure_key() {
  // See above for why we can't use pthread_once.
  pthread_mutex_lock(&mutex);
  static atomic_bool key_created = false;
  if (!key_created) {
    pthread_key_create(&key, run_dtors);
    key_created = true;
  }
  pthread_mutex_unlock(&mutex);
}

int __cxa_thread_atexit_impl(Dtor dtor, void* obj, void* dso_symbol) {
  ensure_key();
  DtorList* old_head = pthread_getspecific(key);
  DtorList* head = (DtorList*)(malloc(sizeof(DtorList)));
  assert(head);
  head->dtor = dtor;
  head->obj = obj;
  head->next = old_head;
  pthread_setspecific(key, head);
  return 0;
}

weak_alias(__cxa_thread_atexit_impl, __cxa_thread_atexit);
