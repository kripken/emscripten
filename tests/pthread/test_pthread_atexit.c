#include <assert.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>

#include "emscripten.h"
#include "emscripten/threading.h"

static _Atomic int threadCounter = 0;
static _Atomic int running = 1;
static pthread_t thread;

void *workerThread(void* arg) {
  threadCounter++;

  while (running)
    emscripten_thread_sleep(1000);

  threadCounter--;

  return NULL;
}

void terminateThread() {
  running = 0;

  int res = 0;
  int rc = pthread_join(thread, (void**)&res);
  assert(rc == 0);
  assert(res == 0);

  printf("done waiting - counter is: %d\n", threadCounter);
}

int main(int argc, char* argv[]) {
  int rc = atexit(terminateThread);
  assert(rc == 0);

  rc = pthread_create(&thread, NULL, workerThread, NULL);
  assert(rc == 0);
  return 0;
}
