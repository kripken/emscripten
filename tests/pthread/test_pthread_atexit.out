done waiting - counter is: 0
